### ISSUE TITLE
The loader might have too many responsibilities

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/testing#5]

**Comment by inf3rno** [comment]
From my perspective it does not seem obvious to make the execution plan with the loader. The loader should load the tests from the modules and deliver them and the dependencies to the executor. I think the executor should load the dependency resolver and decide about the order of execution. Currently it does not make much sense to make the execution plan for the executor with the loader. Another reason not to do that that the series executor can go with a simple array as order while the parallel executor needs to know which tests can run parallel, so it needs 2 collection types: an ordered and an unordered list. If we improve our library further we might need something like executing on multiple threads, comparing performance, and so on. So the order should be decided by the executor and not by the loader.

**Comment by inf3rno** [comment]
I am not sure if we can add some extra info about the latter. So when a test should run parallel with other tests, but I believe we can do it.

**Comment by inf3rno** [comment]
What a loader really needs to do is loading and selecting tests and test groups while excluding tests that does not meet the given criterias entirely. So the executor should know only the tests it must execute and the plan (or selection) should not contain anything that must be excluded. This will be interesting, because the groups can contain both the included and excluded tests and for the reporters we need the groups and something like group completed. Maybe we need to rethink that part too and send only test completed and plan completed events.

**Comment by inf3rno** [comment]
Another solution to have something like `selection.isIncluded(group.firstTest)` and if a test is not included in the selection, then we ignore it in the executor. This would result a lot of unnecessary function calls though.

**Comment by inf3rno** [comment]
I think instead of execution plan I should pass the test selection to the executor. So the selection is not just a query, but it should contain loaded tests. I can do that. I already planned to move the selection building to the loader in #4

**Comment by inf3rno** [comment]
With that in mind we won't need a separate execution plan, at least not outside of the executor. We can add extra properties to our tests if we want to affect execution and the loader won't need to know about them unless we apply a custom filter for them.

**Comment by inf3rno** [comment]
I think we should modify the spec. This will be funny, because I already started to implement it, but I think I can commit just the readme change and nothing else.

**Comment by inf3rno** [comment]
I modified the spec #8 though I did not override the wiki yet. I think this is more about #4 and I wrote down that in our new structure the loader loads the classes and is responsible for finding the tests we want to execute and skip the rest. The executor does the actual executing and builds the report. As of the selector, we should pass a filename and a class from the loader and it should return a true or false, which will mean executing or skipping. I think in this solution the loader is simple enough.
